### PR TITLE
Add Apache Parquet support for TensorFlow Dataset

### PR DESCRIPTION
**Pull request by yongtang**
Apache Parquet is a widely used columnar storage format available in the Hadoop ecosystem. 

This PR is a preliminary attempt to add Apache Parquet support for TensorFlow's Dataset API. It should help many to working on existing parquet-formatted big data with TesnorFlow.

The PR may not cover all the use cases though it could be served as a starting point for further improvement in the future.

The ParquetDataset depends on parquet-cpp (Apache) project as well as other dependencies (e.g, Thrift, etc.). The ParquetDataset only builds on Linux at the moment. This PR also adds the option in `./configure` so that those dependencies could be skipped.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by yongtang** [comment]
@case540 The PR has been updated and build failures have been fixed. Please take a look.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @mrry: It has been 14 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @mrry: It has been 29 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @mrry: It has been 44 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @mrry: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 59 days with no activity and the `awaiting review` label has been applied.

**Comment by mrry** [comment]
I'm terribly sorry for the delay on reviewing this PR. @skye has more experience than I do with the Parquet APIs kindly offered to take over the review. I'll be happy to answer any questions that arise about the `tf.data` integration.

File: tensorflow/contrib/parquet/python/ops/parquet_dataset_ops.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    while True:
+      try:
+        print(sess.run(next_element))
+      except tf.errors.OutOfRangeError:
+        break
+    ```
+
+    Args:
+      filenames: A `tf.string` tensor containing one or more filenames.
+      columns: A 1-D `tf.int32` tensor containing the columns to extract.
```
Comment (skye) [inline tensorflow/contrib/parquet/python/ops/parquet_dataset_ops.py]: nit: I think the kernel allows a scalar or a vector, either fix the docs or make it only accept a vector

File: third_party/parquet/parquet_types.h [hunk 2]
```diff
@@ -0,0 +1,2130 @@
+/**
```
Comment (skye) [inline third_party/parquet/parquet_types.h]: Why do you need to check this in? I'm guessing bazel can't handle building thrift?

**Review by skye — CHANGES_REQUESTED** [review]
Nice! Thanks so much for contributing this!

File: tensorflow/contrib/parquet/BUILD [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        "//third_party/py/numpy",
+        "//tensorflow/python:client_testlib",
+        "//tensorflow/python:framework",
+        "//tensorflow/python:framework_test_lib",
+        "//tensorflow/python:platform_test",
+    ],
+    data = [
+        ":test_data",
+    ],
+    tags = [
```
[the remark this answers is not in the document]
Comment (yongtang) [reply tensorflow/contrib/parquet/BUILD]: Thanks @skye. The tag use `no_pip` as test data needs not to be in pip, also I disabled windows with `no_windows`.

File: third_party/parquet/parquet_types.h (hunk 2 above)
Comment (yongtang) [reply third_party/parquet/parquet_types.h]: Thanks @skye. Parquet needs a thrift type file. This file is generated automatically. This file is also a one time effort for each version and will not change when regenerate. However, it need both flex and bison to build. I tried to put flex and bison in but realized that it is quite complicated. That will adds more dependencies.

On the other hand, as this is just a one time effort of generated file, I think it makes sense to build the file and add the generated file.

**Comment by yongtang** [comment]
Thanks @skye for the review. The comment has been addressed. The current test is failing but that is due to a Bazel bug. I opened the issue in bazel:
https://github.com/bazelbuild/bazel/issues/5932
The bug has been fixed so it will be in the next bazel release.

Can you take a look and see if the PR is OK? Once bazel is updated then all tests should pass I believe.

**Review by skye — DISMISSED** [review]

**Comment by tensorflowbutler** [comment]
Nagging Assignee @case540: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yongtang** [comment]
Bazel 0.17.1 has been released (https://github.com/bazelbuild/bazel/issues/5059#issuecomment-421332163) and it fixed the issue we are facing here. The tests in TF still use Bazel 0.15.0.

A separate PR #22281 has been created to update Bazel to 0.17.1.

/cc @gunan

**Comment by tensorflowbutler** [comment]
Nagging Assignee @case540: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yongtang** [comment]
This PR still depends on #22449 (to update bazel to 0.17.1). Will update the PR once #22449 is fixed.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @case540: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @case540: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @case540: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mrry** [comment]
I'm removing myself as reviewer because I assume that we'll integrate this via SIG IO (where hopefully the same build issues won't be a problem!)...

**Comment by yongtang** [comment]
Since tensorflow/io#21 has been merged. I will close this PR. Thanks everyone for the help!

**Pull request closed without merging**